Bf gint

QingxiangLiu opened the pull request:
# Pull Request Summary
* Fix bugs in ww3_gint.ftn
* Update of mww3_test_03

## Description
This PR fix the bugs in `ww3_gint.ftn` as desribed in #400

### Issue(s) addressed
* Is there an issue associated with this development (bug fix, enhancement, new feature)?    
Closes #400 

### Check list  
* Is your feature branch up to date with the authoritative repository (NOAA/develop)? - YES
* Make sure you have checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop), [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop) and [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) - YES
* Reviewers: @mentions of suggested reviewers of the proposed changes. @aliabdolali @JessicaMeixner-NOAA 


### Testing
* How were these changes tested? - Tested with regtests/mww3_test_03
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) - Not fully, the mww3_test_03 folder is updated slightly.
* Have regression tests been run? - I have only ran regtests/mww3_test_03 and found it works as expected.
* Which compiler / HPC you used to run the regression tests in the PR?  - Intel compiler
* Please provide the summary output of matrix.comp (_matrix.Diff.out_, _matrixCompFull.out_ and _matrixCompSummary.out_):    
Please indicate the expected changes in the outputs ([excluding the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).

Commit:
BF_gint: #400 - fix bugs in ww3_gint.ftn (02/06/2021)

QingxiangLiu committed:
BF_st6: #128 - minor bug in w3src6md (26/05/2021) (#381)

Co-authored-by: Qingxiang Liu <qingxiangl@4000l-145556-l.mobility.unimelb.net.au>

JessicaMeixner-NOAA committed:
update comments and program header for ww3_grid after recent change (#387)

* update comments and program header for ww3_grid after recent change

* fix only statement

Commit:
BF_gint: #400 - update of mww3_test_03 (05/06/2021)

Commit:
BF_gint: #400 - Merge remote-tracking branch 'upstream/develop' into BF_gint

Commit:
BF_gint: #400 - Merge remote-tracking branch 'upstream/develop' into BF_gint

JessicaMeixner-NOAA commented:
@QingxiangLiu does anything need to be added to matix.base so that we are testing this moving forward?

QingxiangLiu commented:
> @QingxiangLiu does anything need to be added to matix.base so that we are testing this moving forward?
Hi Jessica, 

No, it is not very neccessary to update the matrix.base file. But I did update the info file and a few input file for the regtest mww3_test_03 for demonstrating this problem. Please see #400 for more details.

Qingxiang

Commit:
BF_gint: #400 - Merge remote-tracking branch 'upstream/develop' into BF_gint

Commit:
BF_gint: #400 - minor update of input in mww3_test_03

aliabdolali commented:
@QingxiangLiu Thanks for helping to fix this bug. The issue you made and the demonstration with the existing regrests are great. Let me review and run a couple of tests on our end. Your PR is in the queue, so we will get to it soon.

QingxiangLiu commented:
> @QingxiangLiu Thanks for helping to fix this bug. The issue you made and the demonstration with the existing regrests are great. Let me review and run a couple of tests on our end. Your PR is in the queue, so we will get to it soon.

Thank you Ali. Just let me know if anything needs to be done from my side. 

Qingxiang

aliabdolali commented on regtests/mww3_test_03/input/ww3_shel.inp:
```diff
@@ -0,0 +1,29 @@
+$ WAVEWATCH III shell input file
```
mww3_test03 is for multi setup even if it uses one single grid in it. Could you change it ww3_multi.inp and test it.

aliabdolali commented:
I've run the regtests on our hpc and the following tests were different:

```
mww3_test_03/./work_PR3_UQ_MPI_a                     (12 files differ)
mww3_test_03/./work_PR3_UNO_MPI_a                     (12 files differ)
mww3_test_03/./work_PR3_UQ_MPI_a_c                     (12 files differ)
mww3_test_03/./work_PR3_UNO_MPI_a_c                     (12 files differ)
mww3_test_03/./work_PR2_UQ_a                     (12 files differ)
mww3_test_03/./work_PR3_UNO_a_c                     (12 files differ)
mww3_test_03/./work_PR3_UQ_a_c                     (12 files differ)
mww3_test_03/./work_PR1_a                     (12 files differ)
mww3_test_03/./work_PR2_UNO_MPI_a                     (12 files differ)
mww3_test_03/./work_PR3_UNO_a                     (12 files differ)
mww3_test_03/./work_PR2_UNO_a                     (12 files differ)
mww3_test_03/./work_PR1_MPI_a                     (12 files differ)
mww3_test_03/./work_PR2_UQ_MPI_a                     (12 files differ)
mww3_test_03/./work_PR3_UQ_a                     (12 files differ)
```
plus the known non-b4b tests:

```
mww3_test_03
mww3_test_07
ww3_tp2.10
ww3_tp2.15
ww3_tp2.18
```

As I mentioned in a comment, we need to change ww3_shel.inp for mww3_test03 to ww3_multi.inp and then we can merge this PR.

[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6659981/matrixDiff.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6659982/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6659983/matrixCompFull.txt)

Commit:
BF_gint: #400 - minor update of mww3_test_03 (17/06/2021)

Commit:
BF_gint: #400 - Merge remote-tracking branch 'upstream/develop' into BF_gint

QingxiangLiu commented:
> I've run the regtests on our hpc and the following tests were different:
> 
> ```
> mww3_test_03/./work_PR3_UQ_MPI_a                     (12 files differ)
> mww3_test_03/./work_PR3_UNO_MPI_a                     (12 files differ)
> mww3_test_03/./work_PR3_UQ_MPI_a_c                     (12 files differ)
> mww3_test_03/./work_PR3_UNO_MPI_a_c                     (12 files differ)
> mww3_test_03/./work_PR2_UQ_a                     (12 files differ)
> mww3_test_03/./work_PR3_UNO_a_c                     (12 files differ)
> mww3_test_03/./work_PR3_UQ_a_c                     (12 files differ)
> mww3_test_03/./work_PR1_a                     (12 files differ)
> mww3_test_03/./work_PR2_UNO_MPI_a                     (12 files differ)
> mww3_test_03/./work_PR3_UNO_a                     (12 files differ)
> mww3_test_03/./work_PR2_UNO_a                     (12 files differ)
> mww3_test_03/./work_PR1_MPI_a                     (12 files differ)
> mww3_test_03/./work_PR2_UQ_MPI_a                     (12 files differ)
> mww3_test_03/./work_PR3_UQ_a                     (12 files differ)
> ```
> 
> plus the known non-b4b tests:
> 
> ```
> mww3_test_03
> mww3_test_07
> ww3_tp2.10
> ww3_tp2.15
> ww3_tp2.18
> ```
> 
> As I mentioned in a comment, we need to change ww3_shel.inp for mww3_test03 to ww3_multi.inp and then we can merge this PR.
> 
> [matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6659981/matrixDiff.txt)
> [matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6659982/matrixCompSummary.txt)
> [matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6659983/matrixCompFull.txt)

Hi Ali,

Thanks for help on this. 

I just updated the `mww3_test_03` folder by deleting the `ww3_shel.inp` and adding `ww3_multi_grdset_a0.inp`, `grdset_a0` and `ww3_multi_grdset_a0.nml`. The `info` file was also updated.

BTW, to demonstrate the bugs, I added more field parameters for output in `ww3_ounf.inp`. This may affect tests with other grid configurations in mww3_test_03 (e.g., grdset_b/c/d/e) - at least the screen output from ww3_ounf should be affected. But since this was for demonstration only, I can delete the added parameters if necessary (just let me know).

Best regards,
Qingxiang

aliabdolali commented on regtests/mww3_test_03/input/ww3_multi_grdset_a.inp:
```diff
@@ -11,7 +11,8 @@ $
    19680606 000000    600  19680608 000000
      N
-     DPT HS FP DP
+$    DPT HS FP DP
```
could you remove this line?

QingxiangLiu replied:
Done.

aliabdolali commented on regtests/mww3_test_03/input/ww3_ounf.inp:
```diff
@@ -11,7 +11,8 @@ $ file for a full documentation of field output options. Namelist type
 $
   N
-  HS
+$ HS
```
could you remove this line?

QingxiangLiu replied:
Done.

aliabdolali commented:
Hi @QingxiangLiu 
I ran the regtest and we had expected non-b4b due to the changes

```
mww3_test_03/./work_PR3_UQ_MPI_a                     (12 files differ)
mww3_test_03/./work_PR3_UNO_MPI_a                     (12 files differ)
mww3_test_03/./work_PR3_UQ_MPI_a_c                     (12 files differ)
mww3_test_03/./work_PR3_UNO_MPI_a_c                     (12 files differ)
mww3_test_03/./work_PR2_UQ_a                     (12 files differ)
mww3_test_03/./work_PR3_UNO_a_c                     (12 files differ)
mww3_test_03/./work_PR3_UQ_a_c                     (12 files differ)
mww3_test_03/./work_PR1_a                     (12 files differ)
mww3_test_03/./work_PR2_UNO_MPI_a                     (12 files differ)
mww3_test_03/./work_PR3_UNO_a                     (12 files differ)
mww3_test_03/./work_PR2_UNO_a                     (12 files differ)
```

 and preexisting non-b4b tests
```
mww3_test_03
mww3_test_07
ww3_tp2.10
ww3_tp2.15
```
I'll merge the PR once you remove the lines I asked for.

[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6674193/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6674194/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6674195/matrixDiff.txt)

Commit:
BF_gint: #400 - Remove two lines (18/06/2021)

QingxiangLiu commented:
> Hi @QingxiangLiu
> I ran the regtest and we had expected non-b4b due to the changes
> 
> ```
> mww3_test_03/./work_PR3_UQ_MPI_a                     (12 files differ)
> mww3_test_03/./work_PR3_UNO_MPI_a                     (12 files differ)
> mww3_test_03/./work_PR3_UQ_MPI_a_c                     (12 files differ)
> mww3_test_03/./work_PR3_UNO_MPI_a_c                     (12 files differ)
> mww3_test_03/./work_PR2_UQ_a                     (12 files differ)
> mww3_test_03/./work_PR3_UNO_a_c                     (12 files differ)
> mww3_test_03/./work_PR3_UQ_a_c                     (12 files differ)
> mww3_test_03/./work_PR1_a                     (12 files differ)
> mww3_test_03/./work_PR2_UNO_MPI_a                     (12 files differ)
> mww3_test_03/./work_PR3_UNO_a                     (12 files differ)
> mww3_test_03/./work_PR2_UNO_a                     (12 files differ)
> ```
> 
> and preexisting non-b4b tests
> 
> ```
> mww3_test_03
> mww3_test_07
> ww3_tp2.10
> ww3_tp2.15
> ```
> 
> I'll merge the PR once you remove the lines I asked for.
> 
> [matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/6674193/matrixCompFull.txt)
> [matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/6674194/matrixCompSummary.txt)
> [matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/6674195/matrixDiff.txt)

Thanks @aliabdolali. I just delete those two lines.

Qingxiang

aliabdolali commented on model/ftn/ww3_gint.ftn:
```diff
@@ -21,6 +21,7 @@
 !/    31-Aug-2018 : Update groups 2,4,6,8 (S. Zieger)   ( version 6.05 )
 !/    26-Jan-2021 : Added TP field (derived from FP)    ( version 7.12 )
 !/    22-Mar-2021 : New coupling fields output          ( version 7.xx )
+!/    02-Jun-2021 : Big fixs (*SUMGRD; Q. Liu)          ( version 7.xx )
```
typo: Bug fix

QingxiangLiu replied:
Thanks @aliabdolali. I have corrected this typo.

Commit:
BF_gint: #400 - correct typo (20/06/2021)
